### PR TITLE
create wallet

### PR DESCRIPTION
**Pull request by MrHamdulay**

### CODE REVIEW

File: src/config.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+module.exports = config = {
+    PORT: 9000,
```
Comment (MrHamdulay) [inline src/config.js]: this file was not meant to be committed :/ removing

File: src/controllers/wallet.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+		var error_message = '', msisdn, pin;
+        if (req.params) {
+            msisdn = req.params['msisdn'];
+            pin = req.params['pin'];
+        }
+        
+        if (Object.keys(req.params).length === 0) {
+            error_message = 'Please set content-type header approporiately';
+        }
+		else if (pin === undefined ||
```
Comment (MrHamdulay) [inline src/controllers/wallet.js]: updating to make tabs/spaces consistent

**Comment by MrHamdulay** [comment]
@hodgestar @smn @justinvdm ready for review

File: src/controllers/wallet.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
+var WalletModel = require('../models/wallet');
+var Validator = require('validator');
+var Restify = require('restify');
+
+var wallet = {
+    create: function(req, res, next) {
```
Comment (MrHamdulay) [inline src/controllers/wallet.js:15]: hmm, some documentation for this would be nice

File: src/models/wallet.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        var privateKey = key.privateKey;
+        var publicKey = key.publicKey;
+        var encryptionKey = pin;
+        var pinHash = CryptoUtil.hash(pin, salt);
+        var privateKeyEncrypted = CryptoUtil.encryptData(privateKey, encryptionKey);
+        var address = this.addressFromPublicKey(publicKey);
+
+        return DbUtil.promiseConnection()
+            .then(DbUtil.chainedQuery(
+                "INSERT INTO wallet ( msisdn, address, publickey, privatekey, pinhash, salt) VALUES ( $1, $2, $3, $4, $5, $6)",
```
Comment (MrHamdulay) [inline src/models/wallet.js]: this line is waaay too long

File: src/controllers/wallet.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        }
+        
+        if (Object.keys(req.params).length === 0) {
+            error_message = 'Please set content-type header approporiately';
+        }
+        else if (pin === undefined ||
+            msisdn === undefined) {
+            error_message = 'missing argument';
+        }
+        // TODO: this shouldn't just be ZA
```
Comment (smn) [inline src/controllers/wallet.js:31]: I'm happy with this for now. Mobile phone validation that works globally is hard. Perhaps move the `'en-ZA'` to a configurable thing?

File: src/controllers/wallet.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+        if (error_message != '') {
+            return next(new Restify.BadRequestError(error_message));
+        }
+
+        var createWalletPromise = WalletModel.create(msisdn, pin)
+            .done(function(result) {
+                // data from the model is good enough
+                res.send(JSON.stringify(result));
+                next();
```
Comment (smn) [inline src/controllers/wallet.js:51]: What does `next()` do here and wouldn't we want to return the promise from this `create()` function?

File: src/models/wallet.js (hunk 4 above)
Comment (smn) [inline src/models/wallet.js]: Would you consider using [Sequelize](http://docs.sequelizejs.com/en/latest/)?  Looks like a nicely built promises based ORM for Node.js.

File: src/models/wallet.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            .then(function(successResult) {
+                return {
+                    success: true,
+                    publicKey: publicKey,
+                    privateKey: privateKey,
+                    address: address
+                };
+            }, function(failureResult) {
+                return {
+                    success: false
```
Comment (smn) [inline src/models/wallet.js]: We probably want to add something here that gives a hint as to why it failed, I suspect the `failureResult` will have some kind of reason buried in it somewhere?

File: src/utils/crypto.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+var Crypto = {
+    hash: function(data, salt) {
+        salt = salt || '';
+        var subject = data+salt;
+        var hashed = Sjcl.hash.sha256.hash(subject);
+
+        return Sjcl.codec.base64.fromBits(hashed);
+    },
+    // mostly from stellar-wallet-js-sdk (let's not mess with crypto
+    // too much)
```
Comment (smn) [inline src/utils/crypto.js:12]: :D yes.

File: src/wallet.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
+var WalletModel = require('./models/wallet');
+var Validator = require('validator');
+
+var wallet = {
```
Comment (smn) [inline src/wallet.js]: What is this bit of code used for?

**Comment by smn** [comment]
@MrHamdulay initial review done, mostly some minor questions.

File: src/models/wallet.js (hunk 4 above)
Comment (MrHamdulay) [inline src/models/wallet.js]: Will change to Sequelize

**Comment by MrHamdulay** [comment]
updated

File: src/controllers/wallet.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        if (error_message != '') {
+            return next(new Restify.BadRequestError(error_message));
+        }
+
+        var createWalletPromise = WalletModel.create(msisdn, pin)
+            .done(function(result) {
+                // data from the model is good enough
+                res.send(JSON.stringify(result));
+                // calls the next handler in the restify chain
+                next();
```
Comment (smn) [inline src/controllers/wallet.js:51]: Can you explain why we need `next()` here?

**Comment by smn** [comment]
:+1:

File: package.json [hunk 11]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "go-stellar-wallet",
+  "version": "1.0.0",
```
Comment (hodgestar) [inline package.json:3]: Are we confident enough to start at version `1.0.0.`? :)

Comment (MrHamdulay) [inline package.json:3]: I just left all these as default :dancers:

File: src/controllers/wallet.js [hunk 12]
```diff
@@ -0,0 +1,56 @@
+var Validator = require('validator');
+var Restify = require('restify');
+
+var WalletModel = require('../models/wallet');
+var Config = require('../config');
+
+var wallet = {
```
Comment (hodgestar) [inline src/controllers/wallet.js:7]: Perhaps we should rename this to `WalletController` if that's what we're calling it when we require it? :bike:

File: src/controllers/wallet.js [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+     * msisdn - currently limited to one locale (set in Config.VALIDATION_COUNTRY)
+     * pin - integer at least 5 digits
+     */
+    create: function(req, res, next) {
+        // NOTE: request must have content-type set application/json 
+        // or application/x-www-form-urlencoded
+        var error_message = '', msisdn, pin;
+        if (req.params) {
+            msisdn = req.params['msisdn'];
+            pin = req.params['pin'];
```
Comment (hodgestar) [inline src/controllers/wallet.js:21]: Might we worthwhile pulling in something like `lodash` here so we can more easily specify defaults, check for missing keys, etc.

File: src/controllers/wallet.js (hunk 10 above)
Comment (hodgestar) [inline src/controllers/wallet.js:51]: I guess one could also do:

```
.then(function(result) {
   res.send(...);
})
.done(next);
```

but it doesn't feel like a huge win.

Comment (MrHamdulay) [inline src/controllers/wallet.js:51]: I like it

**Pull request merged**